apps/ejabberd/src/shaper.erl

```diff
@@ -1,89 +1,61 @@
-%%%
-%%% This program is distributed in the hope that it will be useful,
-%%% but WITHOUT ANY WARRANTY; without even the implied warranty of
-%%% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-%%% General Public License for more details.
-%%%
-%%% You should have received a copy of the GNU General Public License
-%%% along with this program; if not, write to the Free Software
-%%% Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
-%%% 02111-1307 USA
-%%%
-%%%----------------------------------------------------------------------
+%%==============================================================================
+%% Copyright 2016 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%==============================================================================
 
 -module(shaper).
--author('alexey@process-one.net').
+-author('konrad.zemek@erlang-solutions.com').
 
--export([new/1, new1/1, update/2]).
+-export([new/1, update/2]).
 
--include("ejabberd.hrl").
+-record(shaper, {
+    max_rate :: undefined | pos_integer(),
+    tokens = 0 :: non_neg_integer(),
+    last_update = p1_time_compat:monotonic_time() :: integer()
+}).
 
--record(maxrate, {maxrate, lastrate, lasttime}).
-
--type shaper() :: none | #maxrate{}.
+-type shaper() :: #shaper{}.
 
 -export_type([shaper/0]).
 
 -spec new(atom()) -> shaper().
 new(Name) ->
-    Data = case ejabberd_config:get_global_option({shaper, Name, global}) of
-               undefined ->
-                   none;
-               D ->
-                   D
-           end,
-    new1(Data).
+    case ejabberd_config:get_global_option({shaper, Name, global}) of
+        undefined -> #shaper{};
+        none -> #shaper{};
+        {maxrate, MaxRate} -> #shaper{max_rate = MaxRate, tokens = MaxRate}
+    end.
 
+%% @doc Update shaper.
+%% `Delay' is how many milliseconds to wait.
+-spec update(shaper(), Size :: pos_integer()) -> {shaper(), Delay :: non_neg_integer()}.
+update(#shaper{max_rate = undefined} = Shaper, _Size) ->
+    {Shaper, 0};
+update(Shaper, Size) ->
+    Now = p1_time_compat:monotonic_time(),
+    Second = p1_time_compat:convert_time_unit(1, seconds, native),
 
--spec new1(shaper()) -> shaper().
-new1(none) ->
-    none;
-new1({maxrate, MaxRate}) ->
-    #maxrate{maxrate = MaxRate,
-             lastrate = 0,
-             lasttime = now_to_usec(now())}.
+    TokenGrowth = round(Shaper#shaper.max_rate * (Now - Shaper#shaper.last_update) / Second),
```

Comment:
IMO an expression like ` (Now - Shaper#shaper.last_update) / Second` deserves its own function so that a code reader immediately knowns what it means.